File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> hp = new HashMap<>();
        int max = 0;
        for (int i = 0, j = 0; i < s.length(); i++) {
            if (hp.containsKey(s.charAt(i))) {
                j = Math.max(j, hp.get(s.charAt(i)) + 1);
            }
            hp.put(s.charAt(i), i);
            max = Math.max(max, i + 1 - j);
        }
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int res = 0, left = 0;
        map<char, int> mp;

        for (int right = 0; right < n; right++) {
            if (mp.find(s[right]) != mp.end() && mp[s[right]] >= left) {
                left = mp[s[right]] + 1;
            }
            mp[s[right]] = right;
            res = max(res, right - left + 1);
        }
        
        return res;
    }
};
